File: src/0355_design_twitter.js
/**
 * https://leetcode.com/problems/design-twitter/
 * Your Twitter object will be instantiated and called as such:
 * var obj = new Twitter()
 * obj.postTweet(userId,tweetId)
 * var param_2 = obj.getNewsFeed(userId)
 * obj.follow(followerId,followeeId)
 * obj.unfollow(followerId,followeeId)
 */
class Twitter {
    constructor() {
        this.count = 0;
        this.tweetMap = new Map();
        this.followMap = new Map();
    }

    postTweet(userId, tweetId) {
        if (!this.tweetMap.has(userId)) {
            this.tweetMap.set(userId, []);
        }
        this.tweetMap.get(userId).push([this.count, tweetId]);
        this.count -= 1;
    }

    getNewsFeed(userId) {
        const res = [];
        const minHeap = [];
        if (!this.followMap.has(userId)) {
            this.followMap.set(userId, new Set());
        }
        this.followMap.get(userId).add(userId);

        for (const followeeId of this.followMap.get(userId)) {
            if (this.tweetMap.has(followeeId)) {
                const index = this.tweetMap.get(followeeId).length - 1;
                const [count, tweetId] = this.tweetMap.get(followeeId)[index];
                heappush(minHeap, [count, tweetId, followeeId, index - 1]);
            }
        }

        while (minHeap.length && res.length < 10) {
            const [count, tweetId, followeeId, index] = heappop(minHeap);
            res.push(tweetId);
            if (index >= 0) {
                const [nextCount, nextTweetId] = this.tweetMap.get(followeeId)[index];
                heappush(minHeap, [nextCount, nextTweetId, followeeId, index - 1]);
            }
        }
        return res;
    }

    follow(followerId, followeeId) {
        if (!this.followMap.has(followerId)) {
            this.followMap.set(followerId, new Set());
        }

        this.followMap.get(followerId).add(followeeId);
    }

    unfollow(followerId, followeeId) {
        if (this.followMap.has(followerId)) {
            this.followMap.get(followerId).delete(followeeId);
        }
    }
}

function heapLess(a, b) {
    return a[0] < b[0];
}

function heappush(heap, item) {
    heap.push(item);
    let index = heap.length - 1;
    while (index > 0) {
        const parent = Math.floor((index - 1) / 2);
        if (!heapLess(heap[index], heap[parent])) {
            break;
        }
        [heap[index], heap[parent]] = [heap[parent], heap[index]];
        index = parent;
    }
}

function heappop(heap) {
    const result = heap[0];
    const item = heap.pop();
    if (heap.length > 0) {
        heap[0] = item;
        siftDown(heap, 0);
    }
    return result;
}

function siftDown(heap, index) {
    while (true) {
        const left = index * 2 + 1;
        const right = left + 1;
        let smallest = index;
        if (left < heap.length && heapLess(heap[left], heap[smallest])) {
            smallest = left;
        }
        if (right < heap.length && heapLess(heap[right], heap[smallest])) {
            smallest = right;
        }
        if (smallest === index) {
            break;
        }
        [heap[index], heap[smallest]] = [heap[smallest], heap[index]];
        index = smallest;
    }
}

module.exports = { Twitter };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    let obj = new Twitter();
    obj.postTweet(1, 5);
    assert.deepStrictEqual(obj.getNewsFeed(1), [5]);
    obj.follow(1, 2);
    obj.postTweet(2, 6);
    assert.deepStrictEqual(obj.getNewsFeed(1), [6, 5]);
    obj.unfollow(1, 2);
    assert.deepStrictEqual(obj.getNewsFeed(1), [5]);
}

File: src/0355_design_twitter.rs
use std::cmp::Reverse;
use std::collections::{BinaryHeap, HashMap, HashSet};

struct Twitter {
    count: i32,
    tweet_map: HashMap<i32, Vec<(i32, i32)>>,
    follow_map: HashMap<i32, HashSet<i32>>,
}

impl Twitter {
    fn new() -> Self {
        Self {
            count: 0,
            tweet_map: HashMap::new(),
            follow_map: HashMap::new(),
        }
    }

    fn post_tweet(&mut self, user_id: i32, tweet_id: i32) {
        self.tweet_map
            .entry(user_id)
            .or_default()
            .push((self.count, tweet_id));
        self.count -= 1;
    }

    fn get_news_feed(&mut self, user_id: i32) -> Vec<i32> {
        let mut result = Vec::new();
        let mut heap: BinaryHeap<Reverse<(i32, i32, i32, i32)>> = BinaryHeap::new();
        self.follow_map.entry(user_id).or_default().insert(user_id);

        if let Some(followees) = self.follow_map.get(&user_id) {
            for followee_id in followees {
                if let Some(tweets) = self.tweet_map.get(followee_id) {
                    if let Some(index) = tweets.len().checked_sub(1) {
                        let (count, tweet_id) = tweets[index];
                        heap.push(Reverse((count, tweet_id, *followee_id, index as i32 - 1)));
                    }
                }
            }
        }

        while let Some(Reverse((_, tweet_id, followee_id, index))) = heap.pop() {
            if result.len() == 10 {
                break;
            }
            result.push(tweet_id);
            if index >= 0 {
                if let Some(tweets) = self.tweet_map.get(&followee_id) {
                    let (count, next_tweet_id) = tweets[index as usize];
                    heap.push(Reverse((count, next_tweet_id, followee_id, index - 1)));
                }
            }
        }

        result
    }

    fn follow(&mut self, follower_id: i32, followee_id: i32) {
        self.follow_map
            .entry(follower_id)
            .or_default()
            .insert(followee_id);
    }

    fn unfollow(&mut self, follower_id: i32, followee_id: i32) {
        if let Some(followees) = self.follow_map.get_mut(&follower_id) {
            followees.remove(&followee_id);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        let mut obj = Twitter::new();
        obj.post_tweet(1, 5);
        assert_eq!(obj.get_news_feed(1), vec![5]);
        obj.follow(1, 2);
        obj.post_tweet(2, 6);
        assert_eq!(obj.get_news_feed(1), vec![6, 5]);
        obj.unfollow(1, 2);
        assert_eq!(obj.get_news_feed(1), vec![5]);
    }
}
